### DISCUSSION TITLE
External actions to influence state

### DISCUSSION THREAD

**Discussion by DavidKmn** [pointfreeco/swift-composable-architecture#509] [Q&A] [answered]
Have thoroughly studied the architecture and currently am trying to understand how could an external non user driven effect be managed fed back into the store. 
For example let's say we have a web socket connection where we have events coming in at random non user controlled times. We could take a messaging app as an example where a user receives real time messages via the socket. Presumably we would treat these as actions to affect the state (maybe catch the message data, decode and add it to a dialog), but I'm unclear on how this could be setup using the architecture.

**Comment by stephencelis** [comment, marked as answer]
@DavidKmn This is where the environment comes into play. We have a web socket example here:

https://github.com/pointfreeco/swift-composable-architecture/blob/4449bc2e249ba909086271d37b98f54e0ce725e3/Examples/CaseStudies/SwiftUICaseStudies/02-Effects-WebSocket.swift

It uses a `connectButtonTapped` to kick things off, but you could also tie it to a lifecycle event like `onAppear` or `applicationDidFinishLaunching`.

**Reply by DavidKmn** [reply]
Thanks @stephencelis. Yes this is close to what I was considering. Looking at the example I do have 2 questions. What is the use of the dependencies dictionary (seems like it is used to keep a references) however not sure what its for still. Why are we using a separate reducer for the web socket? should the web socket client not be an environment dependency instead?

**Reply by lukeredpath** [reply]
You don't need to use a dedicate reducer, it's just an example. You would pass the web socket client into whichever domain needs it using the environment and start it from there. 

You may also start it once from say your app domain (e.g. when the app launches) but only send and receive from some child domain within your app.
